rainier-core/src/main/scala/com/stripe/rainier/sampler/LeapFrog.scala

```diff
@@ -1,23 +1,77 @@
 package com.stripe.rainier.sampler
 
-private[sampler] case class LeapFrog(
-    nVars: Int,
-    initialHalfThenFullStep: Array[Double] => Unit,
-    twoFullSteps: Array[Double] => Unit,
-    finalHalfStep: Array[Double] => Unit
-) {
+private[sampler] case class LeapFrog(density: DensityFunction) {
   /*
   Params layout:
   array(0..(n-1)) == ps
   array(n..(n*2-1)) == qs
   array(n*2) == potential
   array(n*2+1) == stepSize
    */
+  private val nVars = density.nVars
   private val potentialIndex = nVars * 2
   private val stepSizeIndex = potentialIndex + 1
   private val inputOutputSize = stepSizeIndex + 1
   private val buf1 = new Array[Double](inputOutputSize)
 
+  def halfPs(inputs: Array[Double]): Unit = {
+    density.update(inputs)
+    var i = 0
+    val j = nVars
+    while (i < j) {
+      inputs(i) = inputs(i) - (inputs(stepSizeIndex) / 2) * density.gradient(i) * -1
```

Comment:
I think `inputs(i) -= ...` is valid syntax here.

Reply:
Updated